File: examples/external-agent-python/agent.py
import json
import os
import time
import urllib.request

BASE_URL = os.environ.get("RELAY_BRIDGE_URL", "http://127.0.0.1:8787")
task = os.environ.get(
    "RELAY_TASK",
    "Review the newest support tickets, flag urgent cases, and draft a summary.",
)

steps = [
    {
        "type": "navigate",
        "title": "Opened support inbox",
        "detail": "Loaded the newest tickets for triage.",
    },
    {
        "type": "read",
        "title": "Read urgent tickets",
        "detail": "Grouped refunds, shipping issues, and escalation requests.",
    },
    {
        "type": "decide",
        "title": "Prepared response plan",
        "detail": "Compiled the next work queue for the operator.",
    },
]


def request(path, payload=None):
    body = None if payload is None else json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        f"{BASE_URL}{path}",
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST" if payload is not None else "GET",
    )

    with urllib.request.urlopen(req) as response:
        raw = response.read().decode("utf-8")
        return json.loads(raw) if raw else None


request("/health")
request("/session/status", {"status": "running"})

for step in steps:
    control = request("/control?consumeRedirect=true")

    if control["redirectInstruction"]:
        task = control["redirectInstruction"]
        request(
            "/action",
            {
                "type": "redirect",
                "title": "Redirect received from Relay",
                "detail": task,
                "reasoning": "Switching to the new operator-provided objective.",
            },
        )

    while control["paused"] and not control["stopRequested"]:
        request("/session/status", {"status": "paused"})
        print("Relay paused the agent. Waiting for resume...")
        time.sleep(1.0)
        control = request("/control?consumeRedirect=true")

    if control["stopRequested"]:
        print("Relay requested stop. Exiting.")
        request("/session/status", {"status": "stopped"})
        raise SystemExit(0)

    request("/session/status", {"status": "running"})
    request(
        "/action",
        {
            **step,
            "reasoning": f"Current task: {task}",
        },
    )
    time.sleep(1.2)

request("/session/status", {"status": "completed"})
print("Demo agent finished successfully.")

File: examples/external-agent-js/agent.mjs
const bridgeUrl = process.env.RELAY_BRIDGE_URL ?? 'http://127.0.0.1:8787'
let task =
  process.env.RELAY_TASK ??
  'Review the newest support tickets, flag urgent cases, and draft a summary.'

const steps = [
  {
    type: 'navigate',
    title: 'Opened support inbox',
    detail: 'Loaded the newest tickets for triage.',
  },
  {
    type: 'read',
    title: 'Read urgent tickets',
    detail: 'Grouped refunds, shipping issues, and escalation requests.',
  },
  {
    type: 'decide',
    title: 'Prepared response plan',
    detail: 'Compiled the next work queue for the operator.',
  },
]

await request('/health')
await request('/session/status', { status: 'running' })

for (const step of steps) {
  let control = await request('/control?consumeRedirect=true')

  if (control.redirectInstruction) {
    task = control.redirectInstruction
    await request('/action', {
      type: 'redirect',
      title: 'Redirect received from Relay',
      detail: task,
      reasoning: 'Switching to the new operator-provided objective.',
    })
  }

  while (control.paused && !control.stopRequested) {
    await request('/session/status', { status: 'paused' })
    console.log('Relay paused the agent. Waiting for resume...')
    await sleep(1000)
    control = await request('/control?consumeRedirect=true')
  }

  if (control.stopRequested) {
    console.log('Relay requested stop. Exiting.')
    await request('/session/status', { status: 'stopped' })
    process.exit(0)
  }

  await request('/session/status', { status: 'running' })
  await request('/action', {
    ...step,
    reasoning: `Current task: ${task}`,
  })
  await sleep(1200)
}

await request('/session/status', { status: 'completed' })
console.log('Demo agent finished successfully.')

async function request(path, payload) {
  const response = await fetch(`${bridgeUrl}${path}`, {
    method: payload ? 'POST' : 'GET',
    headers: { 'Content-Type': 'application/json' },
    body: payload ? JSON.stringify(payload) : undefined,
  })

  const raw = await response.text()
  const data = raw ? JSON.parse(raw) : null

  if (!response.ok) {
    throw new Error(data?.error ?? `Bridge request failed with ${response.status}`)
  }

  return data
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}
